### PR TITLE
Add CMAES

### PR DESCRIPTION
**Pull request by MaxReimann**
CMAES (Covariance Matrix Adaption Evolution Strategy)  is an evolutionary algorithm for difficult non-linear non-convex black-box optimisation problems in continuous domain.

This code is a port of the bare cmaes algorithm in python of Hansen et al.  https://www.lri.fr/~hansen/barecmaes2.html

I think this evolutionary algorithm is a good complement to the existing gradient-based optimization methods in the package.

### CODE REVIEW

**Comment by soumith** [comment]
cc: @fidlej do you want to take a look? I know nothing about CMAES. The code structure of the PR looks fine to me.

File: cmaes.lua [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+-    `state.maxEval`
+        int, maximal number of function evaluations
+-    `state.ftarget`
+        float, target function value
+-    `state.popsize`
+    	 population size. If this is left empty, 
+    	  4 + int(3 * log(|x|)) will be used
+-    `state.ftarget` 
+        stop if fitness < ftarget
+-    `state.verd_disp`
```
Comment (fidlej) [inline cmaes.lua]: It should be verb_disp.

File: doc/index.md [hunk 2]
```diff
@@ -379,3 +380,31 @@ Returns :
...
+Arguments:
+
+* `opfunc` : a function that takes a single input (X), the point of evaluation, and returns f(X) and df/dX. Note that df/dX is not used and can be left 0
+* `x` : the initial point
+* `state.sigma`     : float, initial step-size (standard deviation in each coordinate)
+* `state.maxEval`   : int, maximal number of function evaluations
+* `state.ftarget`   : float, target function value
+* `state.popsize`   : population size. If this is left empty, 4 + int(3 * log(|x|)) will be used
+* `state.ftarget`   : stop if fitness < ftarget
+* `state.verd_disp` : display info on console every verb_disp iteration, 0 for never
```
Comment (fidlej) [inline doc/index.md]: It should be verb_disp.

File: cmaes.lua [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+        int, display on console every verb_disp iteration, 0 for never
+-    `state.args`
```
Comment (fidlej) [inline cmaes.lua]: The opfunc usually takes just `x` as the argument.
If some args need to be passed, the opfunc can be a closure.
Would it be OK to remove the state.objfunc_args usage?

File: cmaes.lua [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+- `x*` : the new `x` vector, at the optimal point
+- `f`  : a table of all function values: 
+     `f[1]` is the value of the function before any optimization and
+     `f[#f]` is the final fully optimized value, at `x*`
+--]]
+function optim.cmaes(opfunc, x, config, state)
+   -- process input parameters
+   local config = config or {}
+   local state = state or config
+   local xmean = torch.Tensor(x:clone():double():storage()) -- distribution mean, a flattened copy
```
Comment (fidlej) [inline cmaes.lua]: It is better to create a flattened view by: x:clone():view(-1)
I guess, a DoubleTensor is not explicitly needed.

File: cmaes.lua [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+   N = xmean:size(1)  -- number of objective variables/problem dimension
```
Comment (fidlej) [inline cmaes.lua]: It should be `local N`.

File: cmaes.lua [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+   local lambda = state.popsize -- population size, offspring number
+   -- Strategy parameter setting: Selection  
+   if state.popsize == nil then
+     lambda = 4 + math.floor(3 * math.log(N))
+   end
+
+   local mu = lambda / 2  -- number of parents/points for recombination
+   local weights = torch.range(0,mu-1):apply(function(i) 
+      return math.log(mu+0.5) - math.log(i+1)  end) -- recombination weights
+    weights:div(weights:sum())  -- normalize recombination weights array
```
Comment (fidlej) [inline cmaes.lua]: The indentation is inconsistent here.

File: cmaes.lua [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+        sigma = sigma * math.exp(math.min(0.6, 
+            (cs / damps) * (torch.sum(torch.pow(ps,2))/N - 1)/2))
+
+      end
+
+      local function stop() 
+        --[[return satisfied termination conditions in a table like 
+        {'termination reason':value, ...}, for example {'tolfun':1e-12}, 
+        or the empty table {}--]] 
+        res = {}
```
Comment (fidlej) [inline cmaes.lua]: Use `local res`.

File: cmaes.lua [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+          print(tostring(counteval).. ': ' .. 
+            string.format(' %6.1f %8.1e ', torch.max(D) / torch.min(D), sigma * max_std) 
+            .. tostring(fitvals[1]))
+        end
+
+        return nil
+      end
+
+      while next(stop()) == nil or iteration < min_iterations do
+        if iterations and iteration >= iterations then
```
Comment (fidlej) [inline cmaes.lua]: `iterations` is undefined. Should be maxEval used here?

File: cmaes.lua [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+      if counteval - eigeneval > lambda/(c1+cmu)/C:size(1)/10 then
+        eigeneval = counteval
+        C = torch.triu(C) + torch.triu(C,1):t() -- enforce symmetry
+        D, B = torch.symeig(C,'V') -- eigen decomposition, B==normalized eigenvectors, O(N^3)
+        D = torch.sqrt(D)  -- D contains standard deviations now
+        invsqrtC = B * torch.diag(torch.pow(D,-1)) * B:t()
+      end
+      res = torch.Tensor(lambda,D:size(1))
+      for k=1,lambda do --repeat lambda times
+        z = D:clone()
```
Comment (fidlej) [inline cmaes.lua]: Use `local res` and `local z`.

File: cmaes.lua [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+            the corresponding objective function values --]]
+    local function tell(arx, _fitvals)
+      -- bookkeeping, preparation
+      counteval = counteval + _fitvals:size(1)  -- slightly artificial to do here
+      N = xmean:size(1)             -- convenience short cuts
+      local iN = torch.range(1,N)
+      local xold = xmean:clone()
+
+      -- Sort by fitness and compute weighted mean into xmean
+      fitvals, arindex = torch.sort(_fitvals)
```
Comment (fidlej) [inline cmaes.lua]: Use `local arindex`.

File: cmaes.lua [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+
+      -- Adapt covariance matrix C
+      local c1a = c1 - (1-hsig^2) * c1 * cc * (2-cc)
+      -- for a minor adjustment to the variance loss by hsig
+      for i=1,N do
+        for j=1,N do
+          local r = torch.range(1,mu)
+          r:apply(function(k) 
+              return weights[k] * (arx[k][i]-xold[i]) * (arx[k][j]-xold[j]) end)
+          Cmuij = torch.sum(r) / sigma^2  -- rank-mu update
```
Comment (fidlej) [inline cmaes.lua]: Use `local Cmuij`.
I'm able to notice the undeclared variables by requiring [https://github.com/deepmind/strict](strict).

File: cmaes.lua [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+          print('termination by', k, '=', v)
+        end
+
+        bestmu, f, c   = best:get()
+        print('best f-value =', f)
+        print('solution = ')
+        print(bestmu)
+        print('best found at iteration: ', c/lambda, ' , total iterations: ', iteration)
+      end
+      table.insert(f_hist, f)
```
Comment (fidlej) [inline cmaes.lua]: The `best:get()` needs to be called to get `f`, if verb_disp is not enabled.

File: cmaes.lua [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+        eigeneval = counteval
+        C = torch.triu(C) + torch.triu(C,1):t() -- enforce symmetry
+        D, B = torch.symeig(C,'V') -- eigen decomposition, B==normalized eigenvectors, O(N^3)
+        D = torch.sqrt(D)  -- D contains standard deviations now
+        invsqrtC = B * torch.diag(torch.pow(D,-1)) * B:t()
+      end
+      res = torch.Tensor(lambda,D:size(1))
+      for k=1,lambda do --repeat lambda times
+        z = D:clone()
+        z:apply(function(d) return d * torch.normal(0,1) end)
```
Comment (fidlej) [inline cmaes.lua]: It would be faster to do:
```
        z = D:clone():normal(0,1):cmul(D)
```

And if you want additional speed improvements, the tensors can be allocated just once.

File: cmaes.lua [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+            string.format(' %6.1f %8.1e ', torch.max(D) / torch.min(D), sigma * max_std) 
+            .. tostring(fitvals[1]))
+        end
+
+        return nil
+      end
+
+      while next(stop()) == nil or iteration < min_iterations do
+        if iterations and iteration >= iterations then
+          return -1
```
Comment (fidlej) [inline cmaes.lua]: The `return -1` does not match the `return bestmu, f_hist` API.

File: cmaes.lua [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+    end
+
+    function BestSolution.update(self, arx, arf, evals)
+      --[[initialize the best solution with `x`, `f`, and `evals`.
+        Better solutions have smaller `f`-values.--]]
+      if self.f == nil or arf < self.f then
+        self.x = arx:clone()
+        self.f = arf
+        if self.evals == nil then
+          self.evals = nil
```
Comment (fidlej) [inline cmaes.lua]: Is this `if` needed? The code would be simpler without it.

**Comment by fidlej** [comment]
Thanks for an interesting black-box optimization algorithm.
I have not studied the correctness of the algorithm.
The test works.

**Comment by fidlej** [comment]
@schaul Do you have some comments?
CMA-ES is similar to NES.
http://schaul.site44.com/publications/nesjmlr.pdf

**Comment by schaul** [comment]
Yes, having implemented CMA-ES a [few times](https://github.com/pybrain/pybrain/blob/master/pybrain/optimization/distributionbased/cmaes.py), I'd say it's an algorithm that's difficult to implement correctly, so I'd suggest a more demanding test (e.g. the 10- or 20-dimensional Rosenbrock instead) and with a tighter budget on evaluations.

File: cmaes.lua (hunk 3 above)
Comment (MaxReimann) [inline cmaes.lua]: yes sure, I removed it.

File: cmaes.lua (hunk 4 above)
Comment (MaxReimann) [inline cmaes.lua]: The problem with using x:clone():view(-1) is, that the Tensor might be of a different type than the default tensor. The internal tensors of cmaes are DoubleTensors by default, which will lead to type mismatch  errors in several operations  (already floatTensors as x don't work). Typing all internal variables as the type of x is also not possible because some Tensors (e.g. cudatensor) don't have all operations implemented. 
Is there a better way to do this ?

File: cmaes.lua (hunk 8 above)
Comment (MaxReimann) [inline cmaes.lua]: Actually iterations was a relict from an earlier version, I removed it entirely.

File: cmaes.lua (hunk 12 above)
Comment (MaxReimann) [inline cmaes.lua]: good catch :)

**Comment by MaxReimann** [comment]
@fidlej Thank you for your thourough review of the code, I implemented nearly all your suggestions. Only line 42 I'm not sure about (see comment).

@schaul The test now uses the 10D Rosenbrock test, running it takes ~40s and 6500 function evals on average.

On a side note, the perfomance is not overwhelming (as in there are CMAES implementations which are faster by factor 5-10). But as most time is spent in the eigenvalue computation, which calls a LAPACK function, there is not much we can do here without making everything much more complicated.

File: cmaes.lua (hunk 4 above)
Comment (fidlej) [inline cmaes.lua]: The other algorithms use `x.new()` to create new tensors. If something is not implemented for CudaTensor, it can be fixed.

Comment (soumith) [inline cmaes.lua]: Yea if something's not implemented for CudaTensor that you need, let me know.

Comment (MaxReimann) [inline cmaes.lua]: Right now, I can only test with cltorch, so I'm not too sure how much is implemented in CudaTensors. What I gathered from looking at the cudatensor docs, torch.diag and torch.triu are not implemented, right ? For cltorch triu, diag and symeig are not implemented.

Comment (fidlej) [inline cmaes.lua]: `torch.diag()` is not implemented for CudaTensor. And `torch.symeig()` requires cutorch compiled with MAGMA.
It seems OK. For now, people can convert `x` to a non-cuda tensor before calling the optimization algorithm.

**Comment by MaxReimann** [comment]
@fidlej I omitted the conversion of the input type to doubletensor and added some checks if the required functions (triu, diag, symeig) are present. If they are implemented in the future for cuda or opencl, the code should also work with no changes for these tensors.

**Comment by fidlej** [comment]
Thanks @MaxReimann . The pull-request looks good.
I don't have the rights to merge it.

**Comment by soumith** [comment]
Thanks for the review Ivo. I gave you write access to optim for the future. Thanks Max for CMAES.
